src/hotspot/share/runtime/thread.cpp

```diff
@@ -1585,157 +1581,109 @@ void JavaThread::remove_monitor_chunk(MonitorChunk* chunk) {
     }
-    // Always null out the _pending_async_exception oop here since the async condition was
-    // already cleared above and thus considered handled.
-    _pending_async_exception = NULL;
-  } else {
-    assert(_is_unsafe_access_error, "must be");
-    DEBUG_ONLY(_is_unsafe_access_error = false);
-
-    // We may be at method entry which requires we save the do-not-unlock flag.
-    UnlockFlagSaver fs(this);
-    Exceptions::throw_unsafe_access_internal_error(this, __FILE__, __LINE__, "a fault occurred in an unsafe memory access operation");
-    // We might have blocked in a ThreadBlockInVM wrapper in the call above so make sure we process pending
-    // suspend requests and object reallocation operations if any since we might be going to Java after this.
-    SafepointMechanism::process_if_requested_with_exit_check(this, true /* check asyncs */);
   }
 }
 
-void JavaThread::handle_special_runtime_exit_condition(bool check_asyncs) {
+void JavaThread::install_async_exception(AsyncExceptionHandshake* aeh) {
+  // Do not throw asynchronous exceptions against the compiler thread.
+  if (!can_call_java()) {
+    delete aeh;
+    return;
+  }
 
-  if (is_obj_deopt_suspend()) {
-    frame_anchor()->make_walkable(this);
-    wait_for_object_deoptimization();
+  // Don't install a new pending async exception if there is already
+  // a pending ThreadDeath one. Just interrupt thread from potential
+  // wait()/sleep()/park() and return.
+  if (has_async_exception_condition(true /* ThreadDeath_only */)) {
+    java_lang_Thread::set_interrupted(threadObj(), true);
+    this->interrupt();
+    delete aeh;
+    return;
   }
 
-  // We might be here for reasons in addition to the self-suspend request
-  // so check for other async requests.
-  if (check_asyncs) {
-    check_and_handle_async_exceptions();
+  oop exception = aeh->exception();
+  Handshake::execute(aeh, this);  // Install asynchronous handshake
+
+  ResourceMark rm;
+  if (log_is_enabled(Info, exceptions)) {
+    log_info(exceptions)("Pending Async. exception installed of type: %s",
+                         InstanceKlass::cast(exception->klass())->external_name());
   }
+  // for AbortVMOnException flag
+  Exceptions::debug_check_abort(exception->klass()->external_name());
 
-  JFR_ONLY(SUSPEND_THREAD_CONDITIONAL(this);)
+  // Interrupt thread so it will wake up from a potential wait()/sleep()/park()
+  java_lang_Thread::set_interrupted(threadObj(), true);
+  this->interrupt();
 }
 
-class InstallAsyncExceptionClosure : public HandshakeClosure {
-  Handle _throwable; // The Throwable thrown at the target Thread
+class InstallAsyncExceptionHandshake : public HandshakeClosure {
+  AsyncExceptionHandshake* _aeh;
 public:
-  InstallAsyncExceptionClosure(Handle throwable) : HandshakeClosure("InstallAsyncException"), _throwable(throwable) {}
-
+  InstallAsyncExceptionHandshake(AsyncExceptionHandshake* aeh) :
+  HandshakeClosure("InstallAsyncException"), _aeh(aeh) {}
```

Comment:
Nit: needs indentation to show it is continuation of previous line.

Reply:
Fixed.